docs/ferite.1
.TH ferite 1L "ferite programming language" "Yeah!" \" -*- nroff -*-

.SH NAME
ferite \- ferite programming lanugage command line tool

.SH SYNOPSIS
\fBferite [options] name_of_script [\-\- options to script]\fP

.SH DESCRIPTION
This little program allows you to run a ferite script from the command line.
It is particularily useful for testing scripts and for writing programs that
are not tied to another application.

.PP
To get a list of options, use 'ferite \-\-help'.
.SH OPTIONS
The
.B ferite
options are listed below.
In keeping with the freedom-of-choice philosophy, options may be eliminated
or default values chosen at compile-time, so options and defaults listed
may not accurately reflect the version installed on your system.

.TP
.BI \-h, \-\-help
Show the current for the program

.TP
.BI \-\-version
Show the current version or the program

.TP
.BI \-\-copyright
Show the current copyright for the program

.TP
.BI \-v, \-\-verbose
This will cause the ferite executable to be verbose, this means that it will
explain where abouts in the grand scheme of things it is, eg. loading a file,
compiling it, deleting it.

.TP
.BI \-s, \-\-showargs
This will cause the program to dump the list of arguments it is passing to
the initialisation of the ferite engine.

.TP
.BI \-e " script" ", \-\-execute "script
This will cause ferite to instead of loading a disk file to compile and run,
use the string
.I script\.
.BR
eg\.
ferite -e "uses 'console'; Console.println('Hello');" - this will print out
.B Hello

.TP
.BI \-c, \-\-compile-only
ferite will only compile the program - this is useful for checking that you
have written syntacically correct code

.TP
.BI \-w, \-\-warnings
ferite will display any warnings that are accumilated whilst compiling or
executing a script.

.TP
.BI \-I "path" ", \-\-include "path
This will add 
.I path
to the list of directories that ferite searches for library
files, this is useful if you  are say testing a library before installing it, or
have particular paths for files [eg. a home directory].

.TP
.BI \-P "module" ", \-\-preload "module
This will add 
.I module
to the list of modules that get loaded into the system before the script is
compiled.

.PP
The following options are specific to the engine and can be used where ever
the engine is used.

.TP
.BI \-\-fe-debug
This will turn on internal debugging of the engine. This will (if the engine
has been compiled with \-\-enable-debug) cause a huge amount of data to be
outputed onto
.I stdout\.
The information contained within it is invaluable - tracking of function
execution flow, all sorts of messages allowing for the very quick location of
bugs.

.TP
.BI \-\-fe-use-classic
ferite by default uses's it's own memory suballocator called
.I jedi
but there are times when it is useful to either access the system's malloc
directly, or to track when memory is being used. If ferite is compiled with
debugging, the classic memory manager will track memory allocations and dump
a table at the end of the program's execution to show all allocated memory.
This is vital for tracking memory leaks.
.I WARNING:
This mode will cause ferite's speed to drop immensily - it's only useful for
debugging the engine.

.TP
.BI \-\-fe-show-mem-use
ferite keeps internal counters of memory useage, use this to see at the end
of a programs execution how much memory is still allocated, or has been
allocated.

.TP
.BI \-\-fe-use-mm-with-pcre
Tell ferite to use it's own memory manager with the pcre regular
expression library

.SH BUGS
Of course! What software project would be without bugs? Please Please, email
bug reports to ferite-devel@lists.sourceforge.net \- thanks. 

.SH AUTHOR
ferite is designed and implemented by Chris Ross <chris@darkrock.co.uk>

.SH URL(s)
ferite Home Page -- http://www.ferite.org/
.br
Author's Home Page -- http://www.darkrock.co.uk/


ChangeLog
-------------------------------------------------------------------------------
15.07.2009 11:30 (BST), ctr

 Make it possible for farm to have a different install location from the 
 prefix of the software install. (for package management)

-------------------------------------------------------------------------------
11.03.2009 12:01 (GMT), ctr

 * Fix some critical bugs in the serialize module for decoding complicated object/array streams

-------------------------------------------------------------------------------
06.03.2009 14:41 (GMT), ctr

 * Compile cache code. When turned on, this will cause ferite to store code in 
   memory as well as code for the disk. This should speed up compiles significantly.
 * It is a lot harder to have duplicate code running around in the source code.
   This covers functions that have the same name and parameter signature.
 * Global cache should make a slight speed improvement.
 * cache_ferite tool for testing cache performance
 * A number of internal bug fixes

-------------------------------------------------------------------------------
15.02.2009 10:08 (GMT), ctr

 * ferite will detect, and use by default, libgc for memory management. It appears
   to work really well with the VM. I even think there might be a speed improvement.
 * Bumped version number to 1.1.12 due to the changes for the GC.

-------------------------------------------------------------------------------
30.09.2008 11:18 (BST), ctr

 * Fix for the error in utf8Length, it was always out by one per character.
 * Added String.nextCharacterLength
 * Added Array.append
 * Bumped version to 1.1.10

-------------------------------------------------------------------------------
12.09.2008 11:09 (BST), ctr

 * Added a 'variables' value to the generate backtrace routine. This captures
   the values of variables and parameters passed into the functions.
 * Fix a potential issue with printing of error messages from error objects
   into the error system.
 * Bumped version to 1.1.9

-------------------------------------------------------------------------------
01.09.2008 12:18 (BST), ctr

 * Fix for a critical ferite memory issue in the error sub-system. Thanks for
   Tobias <tobias.eriksson@cention.se> for spotting the bug.
 * Upped version to 1.1.8

-------------------------------------------------------------------------------
04.07.2008 18:28 (BST), ctr

 * If ferite can't find a directive, it'll now throw an exception rather than
   silently ignore it.

-------------------------------------------------------------------------------
12.06.2008 17:19 (BST), ctr

 * Fix for evil bug that would cause corruption if you accessed a variable
	from a newly created object returned from a function and also did other
	things afterwards e.g.
	
		function a
			return new A();
			
		b( a().id, someOtherfunction() )
		
	The problem was that the return from a() (having not been assigned) was
	tagged as disposable, even though a reference to 'id' was still held. This
	meant that when the internal clean happened, 'id' was killed and we had
	a dangling pointer.

-------------------------------------------------------------------------------
10.06.2008 17:29 (BST), ctr

 * Fixed two critical bugs:

	1) The executor was over keen in freeing variables especially when there is
	variables left on the call record.
	
	2) The is_dots variable was not initialised correctly on a function call 
	record meaning that there is a slight chance that the wrong function would 
	be validated as having the correct parameters.
	
  This was exhibited in random occurrences especially when the jedi memory
  manager is used.

-------------------------------------------------------------------------------
08.04.2008 03:12 (BST), ctr

  * Fix for building executables on Linux

-------------------------------------------------------------------------------
28.03.2008 11:58 (GMT), ctr

 * Make the network module use the correct network read/write functions

-------------------------------------------------------------------------------
04.10.2007 10:52 (BST), ctr

  * Fix directives so that it wont nuke things if you pass too many paramters

-------------------------------------------------------------------------------
29.08.2007 12:42 (BST), ctr

 * Rename ->ccode to ->bytecode
 * Added stuff to allow a backtrace to be obtained at any point within the 
   execution of a program!
 * Obj.backtrace() and Error.backtrace()
 * Moved some functions about to make their location more logical

-------------------------------------------------------------------------------
23.08.2007 12:56 (BST), ctr

 * Fix an error when declaring a namespace function the same name as something 
   else within the same namespace. It would cause memory errors and the like.

-------------------------------------------------------------------------------
23.08.2007 11:29 (BST), ctr

 * You can no longer have empty handle blocks (therefore you can't just suppress
   an error with a monitor{}handle{} block which should reduce sloppy programming
   using this construct)

-------------------------------------------------------------------------------
22.08.2007 12:15 (BST), ctr

 * Documentation for functions in FileSystem.fec, String.fec, Math.fec
 * A couple of functions within the Math namespace for providing a random number
   within a given range and to create a bit array of a given size.
 * feritedoc: @requires allowing the noting of a module an item needs
 * feritedoc: 'provided by' the documentation will note where the class/namespace
   was located.
 * vm: improved error message on backtrace when dealing with functions within a 
   namespace
 * vm: fixed attribute_missing for namespaces

-------------------------------------------------------------------------------
20.08.2007 14:09 (BST), ctr

 * Fix for attribute missing in namespaces

-------------------------------------------------------------------------------
05.07.2007 16:30 (BST), ctr


src/ferite.c
/*
 * Copyright (C) 2000-2007 Chris Ross and various contributors
 * Copyright (C) 1999-2000 Chris Ross
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * o Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 * o Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 * o Neither the name of the ferite software nor the names of its contributors may
 *   be used to endorse or promote products derived from this software without
 *   specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#ifdef HAVE_CONFIG_HEADER
#include "../config.h"
#endif

#include "ferite.h"
#include <math.h>
#include <ferite/fmem_jedi.h>
#include <ferite/fmem_libgc.h>


int ferite_is_initialised = 0;
int ferite_pow_lookup[32];
/**
 * @group General
 * @description These are general functions that are critical to the operation of ferite within 
 *			  another application. This includes the initialisation and deinitialisation of the
 *			  ferite engine aswell as tools to extract the data of parameter lists.
 */

/**
 * @function ferite_init
 * @brief First call to the ferite engine to initialise it
 * @declaration int ferite_init( int argc, char **argv )
 * @param int argc Number of arguments
 * @param char **argc Standard array of pointers
 * @description This must be called to allow the engine to work. Without it, things go
 *			  really rather wrong. You can pass arguments into the scripts and control
 *			  the engine.<nl/>
 *			  <nl/>
 *			  --fe-use-classic - this will tell ferite to use malloc/free rather than the jedi memory manager<nl/>
 *			  --fe-debug - tell ferite to dump debug out to stdout, warning: this will produce a lot of output, ferite also has to be compiled with debugging support.<nl/>
 *			  --fe-show-mem-use - tell ferite to dump to stdout a set of memory statistics, this is useful for detecting leaks<nl/>
 *			  <nl/>
 *			  This function can be called multiple times without fear - it will only set things up
 *			  if they are needed.
 * @return	  1 if ferite engine is already initialised.
 */

int ferite_init( int argc, char **argv )
{
	int i = 0;
	int wantDebugBanner = FE_TRUE;
	
	FE_ENTER_FUNCTION;

	if( !ferite_is_initialised )
	{
		int memory_setup = FE_FALSE;
		int gc_setup = FE_FALSE;

		for( i = 0; i < 32; i++ )
			ferite_pow_lookup[i] = (int)pow( (double)2, (double)i );

		if( argv != NULL )
		{
			for( i = 0; i < argc; i++ )
			{
				if( strcmp( argv[i], "--fe-use-classic" ) == 0 )
				{
					memory_setup = FE_TRUE;
					aphex_init( FE_FALSE );
					ferite_classic_memory_init();
#ifdef DEBUG
					fprintf( stderr, "Using classic memory (system malloc/free)\n");
#endif
				}
				if( strcmp( argv[i], "--fe-use-jedi" ) == 0 )
				{
					memory_setup = FE_TRUE;
					aphex_init( FE_FALSE );
					ferite_jedi_memory_init();
#ifdef DEBUG
					fprintf( stderr, "Using jedi memory\n");
#endif
				}
#ifdef HAVE_LIBGC
				if( strcmp( argv[i], "--fe-use-libgc" ) == 0 )
				{
					memory_setup = FE_TRUE;
					gc_setup = FE_TRUE;
					aphex_init( FE_TRUE );
					ferite_libgc_memory_init();
					ferite_init_gc = ferite_init_libgc_gc;
					ferite_deinit_gc = ferite_deinit_libgc_gc;
					ferite_add_to_gc = ferite_add_to_libgc_gc;
					ferite_check_gc = ferite_check_libgc_gc;
					ferite_merge_gc = ferite_merge_libgc_gc;
#ifdef DEBUG
					fprintf( stderr, "Using libgc memory\n");
#endif
				}
#endif
				if( strcmp( argv[i], "--fe-use-std-gc" ) == 0 )
				{
					gc_setup = FE_TRUE;
					ferite_init_gc = ferite_init_std_gc;
					ferite_deinit_gc = ferite_deinit_std_gc;
					ferite_add_to_gc = ferite_add_to_std_gc;
					ferite_check_gc = ferite_check_std_gc;
					ferite_merge_gc = ferite_merge_std_gc;
				}
				if( strcmp( argv[i], "--fe-use-generation-gc" ) == 0 )
				{
					gc_setup = FE_TRUE;
					ferite_init_gc = ferite_init_generation_gc;
					ferite_deinit_gc = ferite_deinit_generation_gc;
					ferite_add_to_gc = ferite_add_to_generation_gc;
					ferite_check_gc = ferite_check_generation_gc;
					ferite_merge_gc = ferite_merge_generation_gc;
				}
				if( strcmp( argv[i], "--fe-debug" ) == 0 )
				  ferite_show_debug = 1;
				if( strcmp( argv[i], "--fe-show-mem-use" ) == 0 )
				  ferite_hide_mem_use = 0;
				if( strcmp( argv[i], "--fe-use-mm-with-pcre" ) == 0 )
				  ferite_use_mm_with_pcre = 1;
				if( strcmp( argv[i], "--fe-hide-debug" ) == 0 )
					wantDebugBanner = FE_FALSE;
				if( strcmp( argv[i], "--fe-show-partial-implementation") == 0 )
					ferite_show_partial_implementation = FE_TRUE;
			}
		}

		if( !memory_setup ) {
			memory_setup = FE_TRUE;
#ifdef HAVE_LIBGC
			gc_setup = FE_TRUE;
			aphex_init( FE_TRUE );
			ferite_libgc_memory_init();
			ferite_init_gc = ferite_init_libgc_gc;
			ferite_deinit_gc = ferite_deinit_libgc_gc;
			ferite_add_to_gc = ferite_add_to_libgc_gc;
			ferite_check_gc = ferite_check_libgc_gc;
			ferite_merge_gc = ferite_merge_libgc_gc;
#ifdef DEBUG
			fprintf( stderr, "No memory/GC option specified - defaulting to libgc\n");
#endif
#else
			aphex_init( FE_FALSE );
			ferite_jedi_memory_init();
#ifdef DEBUG
			fprintf( stderr, "No memory/GC option specified - defaulting to jedi\n");
#endif
#endif
		}
		if( !gc_setup ) {
			gc_setup = FE_TRUE;
			ferite_init_gc = ferite_init_generation_gc;
			ferite_deinit_gc = ferite_deinit_generation_gc;
			ferite_add_to_gc = ferite_add_to_generation_gc;
			ferite_check_gc = ferite_check_generation_gc;
			ferite_merge_gc = ferite_merge_generation_gc;
		}

#ifdef DEBUG
		if( wantDebugBanner )
			fprintf( stderr, "This a Debug Build.\n" );
#endif

		if( !ferite_init_module_list() )
		{
			fprintf( stderr, "Unable to initialise the module subsystem\n" ); /* FIXME: replace with something nicer */
			ferite_memory_deinit();
			FE_LEAVE_FUNCTION( ferite_is_initialised );
		}

		ferite_init_compiler();
		ferite_init_regex();
		ferite_set_script_argv( 0, NULL );

		ferite_is_initialised = 1;
	}
	FE_LEAVE_FUNCTION( ferite_is_initialised );
}

/**
 * @function ferite_deinit
 * @declaration int ferite_deinit()
 * @brief Shuts the ferite engine down, clears the garbage collector and frees up unused memory
 * @description Use this when the engine is no longer needed. Once this is called you must not call
				any other ferite engine functions. If you wish to re-use ferite after calling this 
				function it is necessary to call ferite_init() again.
 */
int ferite_deinit()
{
	FE_ENTER_FUNCTION;
	if( ferite_is_initialised )
	{
		ferite_variable_destroy( NULL, ferite_ARGV );
		ferite_deinit_module_list();
		ferite_memory_deinit();
		ferite_deinit_regex();
		ferite_deinit_compiler();
		ferite_is_initialised = 0;
	}
	FE_LEAVE_FUNCTION( !ferite_is_initialised );
}

/**
 * @function ferite_get_parameters
 * @declaration int ferite_get_parameters( FeriteVariable **list, int num_args, ... )
 * @brief Get the ferite variables from a list and put them into pointers to real C variables
 * @param FeriteVariable **list The list of variables passed to the function calling this method
 * @param int num_args The number of arguments you wish to extract from the list
 * @param ... Pointers to the real C variables
 * @description This method is very useful for translation from the internal variable format to a standard C variable.
 * @warning Numbers can be either double or long so these need to be handled carefully
 * @example <nl/><code>
	char *str;<nl/>
	FeriteClass *klass;<nl/>
	ferite_get_parameters( params, 2, &str, &klass );<nl/>
   </code>
 */
int ferite_get_parameters( FeriteVariable **list, int num_args, ... )
{
	va_list  ap;
	void	*vptr;
	int	  current_arg = 0;

	FE_ENTER_FUNCTION;

	if( num_args < 1 )
	{
		FE_LEAVE_FUNCTION(1);
	}
	if( list != NULL )
	{
		va_start( ap, num_args );
		while( current_arg < num_args && list[current_arg] != NULL )
		{
			if( list[current_arg] != NULL )
			{
				switch( F_VAR_TYPE(list[current_arg]) )
				{
				  case F_VAR_LONG:
					vptr = va_arg( ap, double * );
					if( vptr == NULL ) break;
					*((double *)vptr) = (long)(list[current_arg]->data.lval);
					break;
				  case F_VAR_DOUBLE:
					vptr = va_arg( ap, double * );
					if( vptr == NULL ) break;
					*((double *)vptr) = list[current_arg]->data.dval;
					break;
				  case F_VAR_STR:
					vptr = va_arg( ap, void ** );
					if( vptr == NULL ) break;
					*((void **)vptr) = list[current_arg]->data.sval;
					break;
				  case F_VAR_OBJ:
				  case F_VAR_UARRAY:
				  case F_VAR_CLASS:
				  case F_VAR_NS:
					vptr = va_arg(ap, void ** );
					if( vptr == NULL ) break;
					*((void **)vptr) = list[current_arg]->data.pval;
					break;
				  case F_VAR_VOID:
					vptr = va_arg(ap, void ** );
					if( vptr == NULL ) break;
					break;
				  case F_VAR_BOOL:
					vptr = va_arg( ap, char * );
					if( vptr == NULL ) break;
					*((char *)vptr) = (char)(list[current_arg]->data.lval);
					break;					
				}
			}
			else
			  break;
			current_arg++;
		}
		va_end( ap );
		FE_LEAVE_FUNCTION( current_arg );
	}
	else
	{
		FE_LEAVE_FUNCTION( -1 );
	}
}

/**
 * @function ferite_get_parameter_count
 * @declaration int ferite_get_parameter_count( FeriteVariable **list )
 * @brief Returns the number of paramters within a list
 * @param FeriteVariable **list The list of variables passed to the function calling this method
 * @description This method is useful for obtaining the number of variables passed to the function. 
				It is suggested that this method is used rather than
 *			  any other as it will allow for underneath changes without breaking existing code.
 */
int ferite_get_parameter_count( FeriteVariable **list )
{
	int argcount = 0;

	FE_ENTER_FUNCTION;
	if( list != NULL )
	{
		while( list[argcount++] != NULL );
		FE_LEAVE_FUNCTION( --argcount );
	}
	else
	{
		FE_LEAVE_FUNCTION( -1 );
	}
}

/**
 * @function ferite_get_required_parameter_list_size
 * @declaration int ferite_get_required_parameter_list_size( FeriteScript *script, FeriteFunction *function )
 * @brief Given a pointer to a function, this will return the expected number of parameters.
 * @param FeriteScript *script The current script
 * @param FeriteFunction *function The function to check
 * @return int The expected count.
 */
int ferite_get_required_parameter_list_size( FeriteScript *script, FeriteFunction *function )
{
   FE_ENTER_FUNCTION;
   FE_LEAVE_FUNCTION( function->arg_count );
}


/**
 * @function ferite_get_parameter
 * @declaration void *ferite_get_parameter( FeriteVariable **list, int num )
 * @brief Get the nth parameter from the parameter list
 * @param FeriteVariable **list The list
 * @param int num The variable to get
 * @description This function will make sure that there isn't an 'out-of-bounds' fault on the parameter array. If you
				are confident of the parameter size it is suggested that you just access the parameter directly.
 */
void *ferite_get_parameter( FeriteVariable **list, int num )
{
	FE_ENTER_FUNCTION;
	if( list != NULL )
	{
		if( num < ferite_get_parameter_count(list) )
		{
			FE_LEAVE_FUNCTION( list[num] );
		}
	}
	FE_LEAVE_FUNCTION( NULL );
}

void ferite_show_help()
{
	FE_ENTER_FUNCTION;
#ifdef DEBUG
	printf( "\nPlease Note: This a Debug Build of ferite\n" );
#endif
	printf( "\nferite specific features:\n\n" );
#ifdef DEBUG
	printf( " --fe-debug	   \t	 Run in debugging-mode.\n" );
#endif
	printf( " --fe-use-classic \t	 Run w/ alternate MM mode. (enables allocation tracking; will cause slow downs)\n" );
	printf( " --fe-use-std-gc  \t	 Run w/ simple GC mode. (will cause slow downs)\n" );
	printf( " --fe-show-mem-use\t	 Report memory use at script end.\n" );
	printf( " --fe-use-mm-with-pcre\t Use PCRE [Regular Expression Engine] with ferite's MM\n" );
	printf( "\n MM = Memory Manager\n" );
	FE_LEAVE_FUNCTION( NOWT );
}

/**
 * @function ferite_set_script_argc
 * @declaration void ferite_set_script_argv( int argc, char **argv )
 * @brief Set the argv array that scripts can access
 * @param int argc The number of arguments
 * @param char **argv The arguments
 * @description The 'argv' array within scripts is used to pass values from outside thee
 *			  script to the inside - much like the way arguments are passed to a C program.
 *			  This function will iterate through the arguments and add them to the globally
				scoped argv variable. When a script is compiled, the array will be copied for
				that script. This means that it is possible to change the arguments to new
				scripts but not existingly compiled scripts.
 */
void ferite_set_script_argv( int argc, char **argv )
{
	int i = 0;
	FeriteScript *script = NULL;
	FE_ENTER_FUNCTION;
	if( ferite_ARGV != NULL )
		ferite_variable_destroy( NULL, ferite_ARGV );

	ferite_ARGV = ferite_create_uarray_variable( NULL, "argv", argc, FE_STATIC );
	if( argv != NULL )
	{
		for( i = 0; i < argc; i++ )
		{
			ferite_uarray_add( NULL, VAUA(ferite_ARGV), fe_new_str(argv[i], argv[i], 0, FE_CHARSET_DEFAULT), NULL, FE_ARRAY_ADD_AT_END );
		}
	}
	FE_LEAVE_FUNCTION(NOWT);
}

/**
 * @end
 */
